### PR TITLE
DCKUBE-621: Prestop hooks 

### PR DESCRIPTION
**Pull request by tarka**
Add prestop hooks to all statefulsets, and normalise the shutdown values for all products.

### CODE REVIEW

**Review by badgersow — APPROVED** [review]

**Comment by badgersow** [comment]
@tarka We can merge this pull request only after all docker images have been updated. Otherwise pre-stop hook is going to fail and the process will be killed immediately by SIGKILL.

**Comment by badgersow** [comment]
As discussed with @errcode1202 , we should also change our `image.pullPolicy` to `Always`

> Always
>    every time the kubelet launches a container, the kubelet queries the container image registry to resolve the name to an image digest. If the kubelet has a container image with that exact digest cached locally, the kubelet uses its cached image; otherwise, the kubelet pulls the image with the resolved digest, and uses that image to launch the container.

Currently if the customers upgrade, their cached image will be old and will not have the shutdown script in the container root. PreStop hook is going to fail and JVM will be killed by SIGKILL

File: src/main/charts/jira/values.yaml [hunk 1]
```diff
@@ -461,6 +461,19 @@ jira:
     #
     secretKey: license-key
 
+  shutdown:
```
Comment (errcode1202) [inline src/main/charts/jira/values.yaml:464]: For consistency can we:

- include a general comment on what this `shutdown` stanza controls 
- leave a carriage return after `shutdown`

File: src/main/charts/bitbucket/values.yaml [hunk 2]
```diff
@@ -567,6 +562,17 @@ bitbucket:
       #
       passwordSecretKey: password
 
+  shutdown:
```
Comment (errcode1202) [inline src/main/charts/bitbucket/values.yaml:565]: For consistency can we:

- include a general comment on what this shutdown stanza controls
- leave a carriage return after shutdown

File: src/main/charts/bitbucket/values.yaml [hunk 3]
```diff
@@ -567,6 +562,17 @@ bitbucket:
...
+    # -- The termination grace period for pods during shutdown. This
+    # should be set to the Bitbucket internal grace period (default 30
+    # seconds), plus a small buffer to allow the JVM to fully terminate.
+    terminationGracePeriodSeconds: 35
+
+    # By default, Bitbucket will be stopped by the script supplied by the image.
+    # If any other shutdown behaviour is needed, it can be achieved by overriding this value.
+    command:
+      - "/shutdown-wait.sh"
```
Comment (errcode1202) [inline src/main/charts/bitbucket/values.yaml:574]: Similar to the `fluentd.command` should we not just default to the `/shutdown-wait.sh` script in the respective helper function, meaning it doesn't need to be explicitly defined here? Looking at the helper for this there isnt actually any need to supply this value here
```
{{/*
The command that will be run to properly shutdown Bitbucket
*/}}
{{- define "bitbucket.shutdown.command" -}}
{{- if .Values.bitbucket.shutdown.command }}
{{- .Values.bitbucket.shutdown.command }}
{{- else }}["/shutdown-wait.sh"]{{- end }}
{{- end }}

```

**Comment by errcode1202** [comment]
> As discussed with @errcode1202 , we should also change the default value of our `image.pullPolicy` to `Always`
> 
> > Always
> > every time the kubelet launches a container, the kubelet queries the container image registry to resolve the name to an image digest. If the kubelet has a container image with that exact digest cached locally, the kubelet uses its cached image; otherwise, the kubelet pulls the image with the resolved digest, and uses that image to launch the container.
> 
> Currently if the customers upgrade, their cached image will be old and will not have the shutdown script in the container root. PreStop hook is going to fail and JVM will be killed by SIGKILL

Based on @nanux latest [comment](https://atlassian.slack.com/archives/GFTNVLQQ4/p1630383517101400?thread_ts=1630380019.091000&cid=GFTNVLQQ4) I think we should leave this for now as is.

File: src/main/charts/bitbucket/values.yaml [hunk 4]
```diff
@@ -567,6 +562,17 @@ bitbucket:
       #
       passwordSecretKey: password
 
+  shutdown:
+    # -- The termination grace period for pods during shutdown. This
+    # should be set to the Bitbucket internal grace period (default 30
+    # seconds), plus a small buffer to allow the JVM to fully terminate.
+    terminationGracePeriodSeconds: 35
+
+    # By default, Bitbucket will be stopped by the script supplied by the image.
```
Comment (errcode1202) [inline src/main/charts/bitbucket/values.yaml:571]: Should we expand on this comment and include a reference/url to what this script is?

**Review by errcode1202 — CHANGES_REQUESTED** [review]

File: src/main/charts/confluence/templates/statefulset.yaml [hunk 5]
```diff
@@ -155,4 +155,4 @@ spec:
         {{ end }}
         {{ include "confluence.volumes" . | nindent 8 }}
         {{ include "fluentd.config.volume" . | nindent 8 }}
-  {{ include "confluence.volumeClaimTemplates" . | nindent 2 }}
\ No newline at end of file
+  {{ include "confluence.volumeClaimTemplates" . | nindent 2 }}
```
[the remark this answers is not in the document]
Comment (badgersow) [reply src/main/charts/confluence/templates/statefulset.yaml:158]: I think that new line was added to the end of the file, hence it appeared on the diff

**Comment by tarka** [comment]
> @tarka We can merge this pull request only after all docker images have been updated. Otherwise pre-stop hook is going to fail and the process will be killed immediately by SIGKILL.

All images have been updated, and I tested that each product has an `/shutdown-wait.sh` and it is being called.

File: src/main/charts/bitbucket/values.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tarka) [reply src/main/charts/bitbucket/values.yaml:574]: The underlying Kubernetes API expects an array of strings (see [ExecAction](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#execaction-v1-core), so I've exposed that.

I'm happy to change this for consistency, but the command needs be wrapped in a shell script, which rapidly descends into quote-hell.

File: src/main/charts/confluence/templates/statefulset.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (tarka) [reply src/main/charts/confluence/templates/statefulset.yaml:158]: This.

File: src/main/charts/bitbucket/values.yaml (hunk 3 above)
Comment (tarka) [reply src/main/charts/bitbucket/values.yaml:574]: We provide many of the default values in the `values.yaml`. To me having the command blank in that context implies no command is run by default, which is not true. We should probably be consistent about this, or add a note about what the default is and where it is set.

File: src/main/charts/bitbucket/templates/_helpers.tpl [hunk 6]
```diff
@@ -123,6 +123,15 @@ The command that should be run by the nfs-fixer init container to correct the pe
 {{- end }}
 {{- end }}
 
+{{/*
+The command that will be run to properly shutdown Bitbucket
+*/}}
+{{- define "bitbucket.shutdown.command" -}}
+{{- if .Values.bitbucket.shutdown.command }}
+{{- .Values.bitbucket.shutdown.command }}
+{{- else }}["/shutdown-wait.sh"]{{- end }}
```
[the remark this answers is not in the document]
Comment (tarka) [reply src/main/charts/bitbucket/templates/_helpers.tpl:132]: This goes to @errcode1202 's comment below about default values. I lean to having them in the `values.yaml` for clarity, but we should be consistent.

File: src/main/charts/bitbucket/values.yaml (hunk 4 above)
Comment (tarka) [reply src/main/charts/bitbucket/values.yaml]: Done

File: src/main/charts/crowd/values.yaml [hunk 7]
```diff
@@ -213,6 +213,22 @@ crowd:
     #
     usePodNameAsClusterNodeName: true
 
+  shutdown:
+
+    # -- The termination grace period for pods during shutdown. This
+    # should be set to the internal grace period, plus a small buffer
+    # to allow the JVM to fully terminate.
```
Comment (errcode1202) [inline src/main/charts/crowd/values.yaml:220]: nit - for consistency can we include a `#` after the the property name for all of these i.e.

```
# -- The termination grace period for pods during shutdown. This
# should be set to the internal grace period, plus a small buffer
# to allow the JVM to fully terminate.
#
terminationGracePeriodSeconds: 30
```

**Review by errcode1202 — CHANGES_REQUESTED** [review]

File: src/main/charts/crowd/values.yaml (hunk 7 above)
Comment (tarka) [reply src/main/charts/crowd/values.yaml:220]: Done

**Review by nanux — APPROVED** [review]

**Review by errcode1202 — APPROVED** [review]
🍻

**Review by nghazali — APPROVED** [review]

**Pull request merged**